File: Binary Search/BS on Answers/find min:max/Smallest_divisor_threshold.java
public class Smallest_divisor_threshold{
    public static void main(String[] args){
        int[] arr = {1, 2, 5, 9};
        int threshold = 6;

        int low = 0, high = get_max(arr);
        int ans = 0;

        while (low <= high) { 
            int mid = (low + high)/2;

            if (sumOfD(arr, threshold, mid) <= threshold) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        System.out.println("Smallest divisor is : " + ans);
    }

    public static int get_max(int[] arr){
        int max = Integer.MIN_VALUE;

        for (int i = 0; i < arr.length; i++) {
            if(arr[i] > max){
                max = arr[i];
            }
        }
        return max;
    }

    public static int sumOfD(int[] arr, int threshold, int mid){
        int sum = 0;

        for(int i = 0; i < arr.length; i++){
            sum += (arr[i] + mid - 1)/mid;
        }
        return sum;
    }
}

File: Binary Search/BS on Answers/Python/find minmax/Smallest_divisor_threshold.py
import math

def smallestDivisor(nums, threshold):
    low, high = 1, max(nums)

    res = high

    while low <= high:
        mid = (low + high) // 2

        sum = 0
        for num in nums:
            sum += math.ceil(num / mid)
        
        if sum <= threshold:
            res = mid
            high = mid - 1
        else:
            low = mid + 1

    return res

nums = [44,22,33,11,1]
threshold = 5
print(smallestDivisor(nums, threshold))